edge/pkg/devicetwin/dtmanager/communicate_test.go

```diff
@@ -0,0 +1,336 @@
+/*
+Copyright 2019 The KubeEdge Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dtmanager
+
+import (
+	"errors"
+	"github.com/kubeedge/kubeedge/common/beehive/pkg/core/context"
+	"github.com/kubeedge/kubeedge/common/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcontext"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dttype"
+	"reflect"
+	"testing"
+	"time"
+)
+
+// TestStartAction is function to test Start() when value is passed in ReceiverChan.
+func TestStartAction(t *testing.T) {
+	mainContext := context.GetContext(context.MsgCtxTypeChannel)
+	dtContextStateConnected, _ := dtcontext.InitDTContext(mainContext)
+	dtContextStateConnected.State = dtcommon.Connected
+	receiveChanActionPresent := make(chan interface{}, 1)
+	receiveChanActionPresent <- &dttype.DTMessage{Action: dtcommon.SendToCloud, Identity: "identity", Msg: &model.Message{Header: model.MessageHeader{ID: "message"}, Content: "msg"}}
+	receiveChanActionNotPresent := make(chan interface{}, 1)
+	receiveChanActionNotPresent <- &dttype.DTMessage{Action: "action", Identity: "identity", Msg: &model.Message{Content: "msg"}}
+	tests := []struct {
+		name   string
+		Worker Worker
+		Group  string
+	}{
+		{
+			name: "StartTest-ActionPresentInActionCallback",
+			Worker: Worker{
+				ReceiverChan: receiveChanActionPresent,
+				DTContexts:   dtContextStateConnected,
+			},
+			Group: "",
+		},
+		{
+			name: "StartTest-ActionNotPresentInActionCallback",
+			Worker: Worker{
+				ReceiverChan: receiveChanActionNotPresent,
+				DTContexts:   dtContextStateConnected,
+			},
+			Group: "",
+		},
+	}
+	for _, test := range tests {
+		cw := CommWorker{
+			Worker: test.Worker,
+			Group:  test.Group,
+		}
+		go cw.Start()
+		time.Sleep(1 * time.Millisecond)
```

Comment:
can we keep the 3 steps above inside t.Run ??
applicable to all similar cases below

Reply:
Updated